include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h

```diff
@@ -0,0 +1,547 @@
+    ::std::size_t __segments =
+        oneapi::dpl::__internal::__dpl_ceiling_div(__n, __work_group_size * __iters_per_work_item);
+
+    auto __private_histograms =
+        oneapi::dpl::__par_backend_hetero::__buffer<_ExecutionPolicy, _bin_type>(__exec, __segments * __num_bins)
+            .get_buffer();
+
+    return __exec.queue().submit([&](auto& __h) {
+        __h.depends_on(__init_e);
+        oneapi::dpl::__ranges::__require_access(__h, __input, __bins, __opt_range...);
+        sycl::accessor __hacc_private{__private_histograms, __h, sycl::read_write, sycl::no_init};
+        __h.parallel_for(
+            sycl::nd_range<1>(__segments * __work_group_size, __work_group_size), [=](sycl::nd_item<1> __self_item) {
+                constexpr auto _atomic_address_space = sycl::access::address_space::global_space;
+                const ::std::size_t __self_lidx = __self_item.get_local_id(0);
+                const ::std::size_t __wgroup_idx = __self_item.get_group(0);
+                const ::std::size_t __seg_start = __work_group_size * __iters_per_work_item * __wgroup_idx;
+
+                __clear_wglocal_histograms(__hacc_private, __wgroup_idx * __num_bins, __num_bins, __self_item);
+
+                if (__seg_start + __work_group_size * __iters_per_work_item < __n)
+                {
+                    for (::std::size_t __idx = 0; __idx < __iters_per_work_item; __idx++)
+                    {
+                        ::std::size_t __val_idx = __seg_start + __idx * __work_group_size + __self_lidx;
+                        __accum_local_atomics_iter<_histogram_index_type, _atomic_address_space>(
+                            __input, __val_idx, __hacc_private, __wgroup_idx * __num_bins, __func);
+                    }
+                }
+                else
+                {
+                    for (::std::size_t __idx = 0; __idx < __iters_per_work_item; __idx++)
+                    {
+                        ::std::size_t __val_idx = __seg_start + __idx * __work_group_size + __self_lidx;
+                        if (__val_idx < __n)
+                        {
+                            __accum_local_atomics_iter<_histogram_index_type, _atomic_address_space>(
+                                __input, __val_idx, __hacc_private, __wgroup_idx * __num_bins, __func);
+                        }
+                    }
+                }
+
+                __dpl_sycl::__group_barrier(__self_item);
+
+                __reduce_out_histograms<_bin_type>(__hacc_private, __wgroup_idx * __num_bins, __bins, __num_bins,
+                                                   __self_item);
+            });
+    });
+}
+
+template <::std::uint16_t __iters_per_work_item, typename _ExecutionPolicy, typename _Iter1, typename _Iter2,
+          typename _Size, typename _IdxHashFunc, typename... _Range>
+inline auto
+__parallel_histogram_sycl_impl(_ExecutionPolicy&& __exec, _Iter1 __first, _Iter1 __last, _Iter2 __histogram_first,
```

Comment:
1. we already mentioned "sycl" in the filename, so probably we dont need duplicate it in the name of the function(?)
2. Also I'm confused with argument types - there are "iterators" and "ranges" at the same time... Before we have or an iterator based signature or a range based signature for an algorithm.

Reply:
1.  OK, changed the name.

2.  The public API is iterators only.  Previously, this code location was in the middle of conversion of different inputs to sycl ranges, which occurred at different times because of the unification of the 2 APIs.  
    
	To make this more consistent, I moved the sycl range transformation for the always required input elements and and output bins up the stack to `__parallel_histogram`, which makes each level either all ranges or all iterators, and still allows it to be unified for both APIs. Thanks, this refactor improved the code.